op-batcher/driver.go

```diff
@@ -170,147 +168,83 @@ func (l *BatchSubmitter) Stop() {
 	l.wg.Wait()
 }
 
+// loadBlocksIntoState loads all blocks since the previous submitted block
+func (l *BatchSubmitter) loadBlocksIntoState(ctx context.Context) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	syncStatus, err := l.cfg.RollupNode.SyncStatus(ctx)
+	if err != nil {
+		l.log.Warn("issue fetching L2 head", "err", err)
+		return
+	}
+	if syncStatus.HeadL1 == (eth.L1BlockRef{}) {
+		l.log.Info("Rollup node has no L1 head info yet")
+		return
+	}
+	l.log.Info("Got new L2 sync status", "safe_head", syncStatus.SafeL2, "unsafe_head", syncStatus.UnsafeL2, "last_submitted", l.lastSubmittedBlock, "l1_head", syncStatus.HeadL1)
+	if syncStatus.SafeL2.Number >= syncStatus.UnsafeL2.Number {
+		l.log.Trace("No unsubmitted blocks from sequencer")
+		return
+	}
+
+	// If we just started, start at safe-head
+	if l.lastSubmittedBlock == (eth.BlockID{}) {
+		l.log.Info("Starting batch-submitter work at safe-head", "safe", syncStatus.SafeL2)
+		l.lastSubmittedBlock = syncStatus.SafeL2.ID()
+	}
+	// If it's lagging behind, catch it up.
+	if l.lastSubmittedBlock.Number < syncStatus.SafeL2.Number {
+		l.log.Warn("last submitted block lagged behind L2 safe head: batch submission will continue from the safe head now", "last", l.lastSubmittedBlock, "safe", syncStatus.SafeL2)
+		l.lastSubmittedBlock = syncStatus.SafeL2.ID()
+	}
+
+	prevID := l.lastSubmittedBlock
+
+	// Add all blocks to "state"
+	for i := l.lastSubmittedBlock.Number + 1; i <= syncStatus.UnsafeL2.Number; i++ {
+		ctx, cancel := context.WithTimeout(l.ctx, time.Second*10)
+		block, err := l.cfg.L2Client.BlockByNumber(ctx, new(big.Int).SetUint64(i))
+		cancel()
+		if err != nil {
+			l.log.Error("issue fetching L2 block", "err", err)
+			return
+		}
+		if block.ParentHash() != prevID.Hash {
+			l.log.Error("detected a reorg in L2 chain vs previous submitted information, resetting to safe head now", "safe_head", syncStatus.SafeL2)
+			l.lastSubmittedBlock = syncStatus.SafeL2.ID()
+			return
+		}
+		if err := l.state.AddL2Block(block); err != nil {
+			return
+		}
+		prevID = eth.ToBlockID(block)
+		l.lastSubmittedBlock = prevID
```

Comment:
This line is new (it would previously only update it when successfully submitting a tx for it)? We may want to rename the variable to last-buffered to make sure we still submit data for this.

Reply:
makes sense. The channel manager is going to be more responsible for resubmitting data in the future.